### ISSUE TITLE
Using python27 with future results in TypeError: Cannot mix str and non-str arguments

### ISSUE THREAD

**Issue by roemhildtg** [jfinkels/flask-restless#594]
![image](https://cloud.githubusercontent.com/assets/5471079/19243633/ee978dba-8edd-11e6-9b28-c1a5cd62ceed.png)

I've written some code in python 3 but decided to revert back to using python 27 for compatibility with some other libraries, and to increase compatibility with the code I've written am using the future library. 

I'm not too certain what the cause is but it looks like we're [passing a Unicode and a String](https://github.com/jfinkels/flask-restless/blob/master/flask_restless/serialization/serializers.py#L382) to urllib which apparently is a problem in the future library.

**Comment by roemhildtg** [comment]
This can be closed, I switched to using the six library and now its working. 
```python
from six.moves.urllib.request import pathname2url
```

Previously I was doing this in a view:
```python
from future.standard_library import install_aliases
install_aliases()
```

**Comment by jfinkels** [comment]
Before you close this, is there something we can do in Flask-Restless to mitigate this issue? The non-unicode string seems to come from `flask_restless.url_for`, which in turn comes from `flask.url_for`.

**Comment by roemhildtg** [comment]
I'm not really sure how we could fix it. I think the problem is that [request.url_root](http://flask.pocoo.org/docs/0.11/api/#flask.Request.url_root) is unicode and url_for returns a different type of string (I apologize, I'm not sure what exactly, but it was something like future.types.str). 

Then urllib complains because the two types are not the same type of instance.

**Comment by roemhildtg** [comment]
Yeah, I went back and took a double look:
The two strings being passed here are `<type unicode>` and `<type 'future.types.newstr.newstr'>`

The future lib calls this function with those arguments: 
```python
def _coerce_args(*args):
    # Invokes decode if necessary to create str args
    # and returns the coerced inputs along with
    # an appropriate result coercion function
    #   - noop for str inputs
    #   - encoding function otherwise
    str_input = isinstance(args[0], str)
    for arg in args[1:]:
        # We special-case the empty string to support the
        # "scheme=''" default argument to some functions
        if arg and isinstance(arg, str) != str_input:
            raise TypeError("Cannot mix str and non-str arguments")
    if str_input:
        return args + (_noop,)
    return _decode_args(args) + (_encode_result,)
```

**Comment by roemhildtg** [comment]
So I guess we could wrap the unicode in a `str(url)` perhaps?

**Comment by jfinkels** [comment]
I think for future-proofing, I'd like to promote both arguments to unicode strings instead of demote both arguments to (Python 2) strings. I'll do something like
```python
                url = urljoin(request.url_root, to_unicode(path))
```
and see if that works. Can I reproduce the problem simply by doing
```python
from future.standard_library import install_aliases
install_aliases()
```
in my unit tests, for example?

**Comment by roemhildtg** [comment]
Yes, that's all I did to produce the error. It should be all you need to do. The error is thrown when a request is made to the api.

**Comment by jfinkels** [comment]
I've added a fix for this in pull request #599, which is currently building. Let's see if it passes!

**Comment by jfinkels** [comment]
Fixed by pull request #599.
